File: output/lec-python-redo/course/lesson10/in_class/practice06.py
"""
[难度: ★★★★][所属知识点: 综合(完整购物车系统)][预计完成时间: 30 分钟]

题目:把 Day 10 所有函数组合成完整的购物车系统。
要求:
- 商品库至少 4 个商品
- 菜单循环(1.浏览 2.加购 3.查看 4.结算 0.退出)
- 所有 input 用 try-except 包裹
- 结算后清空购物车
"""

if __name__ == "__main__":
    print("参考答案:")
    print('import json, os')
    print('')
    print('DATA_FILE = "products.json"')
    print('products = [')
    print('    {"id": 1, "name": "Python 入门", "price": 39.9},')
    print('    {"id": 2, "name": "NumPy 实战",   "price": 49.9},')
    print('    {"id": 3, "name": "Pandas 精讲",  "price": 59.9},')
    print('    {"id": 4, "name": "爬虫基础",     "price": 29.9},')
    print(']')
    print('cart = []')
    print('')
    print('def load_products(): ...')
    print('def show_menu(): ...')
    print('def show_products(): ...')
    print('def add_to_cart(): ...')
    print('def view_cart(): ...')
    print('def checkout(): ...')
    print('def save_products(): ...')
    print('')
    print('def main():')
    print('    load_products()')
    print('    while True:')
    print('        choice = show_menu()')
    print('        if choice == "1": show_products()')
    print('        elif choice == "2": add_to_cart()')
    print('        elif choice == "3": view_cart()')
    print('        elif choice == "4": checkout()')
    print('        elif choice == "0":')
    print('            save_products()')
    print('            break')
    print('')
    print('if __name__ == "__main__":')
    print('    main()')

File: output/iter-R9-js-build-pedagogy/course/lesson02/in_class/practice06.js
/*
[难度: ★★★]
[所属知识点: 综合(挑战题)]
[预计完成时间: 15 分钟]
[类型: 挑战题,不强制完成]

题目:下面的"购物车"代码有 4 处反模式,找出来并修正。

    function checkout(cart) {
        var total = 0;
        for (var i = 0; i < cart.length; i++) {
            var price = cart[i].price;
            var tax = price * 0.1;
            total = total + price + tax;
        }

        var DISCOUNT = 0.9;
        if (total > 100) {
            DISCOUNT = 0.8;
        }
        total = total * DISCOUNT;

        return total;
    }

反模式提示:var 泄漏 / const 误用 / 作用域 / 命名规范
*/

if (require.main === module) {
  console.log("4 处反模式:");
  console.log("  1. var i 在 for 循环外泄漏 → 修正:let i");
  console.log("  2. var price/tax 在 for 块外泄漏 → 修正:let price/let tax");
  console.log("  3. var DISCOUNT 重新赋值 → 修正:let DISCOUNT(因为会变)");
  console.log("     (或 const + 三元运算符)");
  console.log("  4. var total 可以,const 不行(因为重新赋值) → 保持 let");
  console.log("");
  console.log("修正后:");
  console.log("  function checkout(cart) {");
  console.log("      let total = 0;");
  console.log("      for (let i = 0; i < cart.length; i++) {");
  console.log("          let price = cart[i].price;");
  console.log("          let tax = price * 0.1;");
  console.log("          total = total + price + tax;");
  console.log("      }");
  console.log("      let DISCOUNT = total > 100 ? 0.8 : 0.9;");
  console.log("      total = total * DISCOUNT;");
  console.log("      return total;");
  console.log("  }");
}
